mm2src/common/executor/abortable_system/abortable_queue.rs

```diff
@@ -0,0 +1,274 @@
+                // to complete in `critical_timeout_s` seconds.
+                if let Some(critical_timeout_s) = critical_timeout_s {
+                    Timer::sleep(critical_timeout_s).await;
+                }
+            };
+
+            match select(abortable_fut.boxed(), wait_till_abort.boxed()).await {
+                // The future has finished normally.
+                Either::Left(_) => {
+                    if let Some(on_finish) = settings.on_finish {
+                        log::log!(on_finish.level, "{}", on_finish.msg);
+                    }
+
+                    // We need to remove the future ID if the handler still exists.
+                    if let Some(inner) = inner_weak.upgrade() {
+                        inner.lock().remove_finished(future_id);
+                    }
+                },
+                // `abort_tx` has been removed from `QueueInner::futures`,
+                // *and* the `critical_timeout_s` timeout has expired (if was specified).
+                Either::Right(_) => {
+                    if let Some(on_abort) = settings.on_abort {
+                        log::log!(on_abort.level, "{}", on_abort.msg);
+                    }
+
+                    // Abort the input `fut`.
+                    abort_handle.abort();
+                },
+            }
+        };
+
+        unsafe { spawn(final_fut) };
+    }
+}
+
+/// `QueueInner` is the container of the spawned future handles [`oneshot::Sender<()>`].
+/// It holds the future handles, gives every future its *unique* `FutureId` identifier
+/// (unique between spawned and alive futures).
+/// Once a future is finished, its `FutureId` can be reassign to another future.
+/// This is necessary so that this container does not grow indefinitely.
+pub struct QueueInner {
+    abort_handlers: Vec<oneshot::Sender<()>>,
+    finished_futures: Vec<FutureId>,
+}
+
+impl Default for QueueInner {
+    fn default() -> Self {
+        QueueInner {
+            abort_handlers: Vec::with_capacity(CAPACITY),
+            finished_futures: Vec::with_capacity(CAPACITY),
+        }
+    }
+}
+
+impl QueueInner {
+    /// Inserts the given future `handle`.
+    fn insert_handle(&mut self, handle: oneshot::Sender<()>) -> FutureId {
+        match self.finished_futures.pop() {
+            Some(finished_id) => {
+                self.abort_handlers[finished_id] = handle;
+                // The freed future ID.
+                finished_id
+            },
+            None => {
+                self.abort_handlers.push(handle);
+                // The last item ID.
+                self.abort_handlers.len() - 1
+            },
+        }
+    }
+
+    /// # Note
+    ///
+    /// We don't need to remove an associated [`oneshot::Sender<()>`],
+    /// but later we can easily reset the item at `abort_handlers[future_id]` with a new [`oneshot::Sender<()>`].
+    fn remove_finished(&mut self, future_id: FutureId) { self.finished_futures.push(future_id); }
```

Comment:
The method is called `remove_finished`, but it actually pushes a new element to `finished_futures`. Could you please try choosing a more relevant one?

Reply:
Renamed to `on_finished`